R/data.table.R

```diff
@@ -1385,7 +1385,8 @@ replace_dot_alias = function(e) {
     byval = i
     bynames = if (missing(on)) head(key(x),length(leftcols)) else names(on)
     allbyvars = NULL
-    bysameorder = haskey(i) || (is.sorted(f__) && ((roll == FALSE) || length(f__) == 1L)) # Fix for #1010
```

Comment:
I removed the `length(f__)==1L` too. That was there, iiuc, from before https://github.com/Rdatatable/data.table/pull/3443 when `is.sorted()` returned false for a single NA.  Since `is.sorted()` is true for all length 1 input, the `|| length(f__)==1L` is superfluous now.
Also moved the `roll=FALSE` first before `&& is.sorted(f__)` to save the call to `is.sorted()` when `roll==TRUE`.

Reply:
`chmatch` not `match` too. i can fix if you're AFK now.

Reply:
good spot. done.

Reply:
i thought we had a `isLeadingSubsetOfKey()` helper function, or similar name, as we do that op quite a bit.  But when I did a few greps I couldn't find anything.

Reply:
I was exactly thinking that we should just have a helper for this by now...